File: booking-service/src/main/java/entelect/training/incubator/spring/booking/config/RewardsConfig.java
package entelect.training.incubator.spring.booking.config;

import entelect.training.incubator.spring.booking.api.RewardsApiClient;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.oxm.jaxb.Jaxb2Marshaller;

@Configuration
@ConditionalOnProperty(
        name = "rewards.endpoint",
        matchIfMissing = false
)
public class RewardsConfig {

    @Value("${rewards.endpoint}")
    private String rewardsEndpoint;

    @Bean
    public Jaxb2Marshaller marshaller() {
        Jaxb2Marshaller marshaller = new Jaxb2Marshaller();
        marshaller.setContextPath("com.baeldung.springsoap.client.gen");
        return marshaller;
    }

    @Bean
    public RewardsApiClient rewardsClient(Jaxb2Marshaller marshaller) {
        RewardsApiClient client = new RewardsApiClient();
        client.setDefaultUri(rewardsEndpoint); // ✅ dynamic, container-safe
        client.setMarshaller(marshaller);
        client.setUnmarshaller(marshaller);
        return client;
    }
}

File: booking-service/src/main/kotlin/entelect/training/incubator/spring/booking/config/RewardsConfig.kt
package entelect.training.incubator.spring.booking.config

import entelect.training.incubator.spring.booking.api.RewardsApiClient
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.oxm.jaxb.Jaxb2Marshaller


@Configuration
@ConditionalOnProperty(name = ["rewards.endpoint"], matchIfMissing = false)
class RewardsConfig {
    @Value("\${rewards.endpoint}")
    private val rewardsEndpoint: String? = null

    @Bean
    fun marshaller(): Jaxb2Marshaller {
        val marshaller = Jaxb2Marshaller()
        marshaller.setContextPath("com.baeldung.springsoap.client.gen")
        return marshaller
    }

    @Bean
    fun rewardsClient(marshaller: Jaxb2Marshaller?): RewardsApiClient {
        val client: RewardsApiClient = RewardsApiClient()
        client.setDefaultUri(rewardsEndpoint)
        client.setMarshaller(marshaller)
        client.setUnmarshaller(marshaller)
        return client
    }
}